src/matmul_kernels.py

```diff
@@ -0,0 +1,157 @@
+    pid = tl.program_id(axis=0)  # [0, 1, ... {out_features-1}]
+
+    # Stage 1: load input data
+    input_vec = tl.load(
+        input_vec_ptr
+        + tl.arange(0, num_input_groups_next_power_of_2)[:, None, None] * in_group_size
+        + tl.arange(0, in_group_size)[None, None, :],
+        mask=tl.arange(0, num_input_groups_next_power_of_2)[:, None, None] < num_input_groups,
+    )
+    # [in_features//in_group_size, 1, group_size]
+    # Note: we could simply load input_vec then reshape
+    #     input_vec = tl.load(input_vec_ptr + tl.arange(0, in_features))  # [in_features]
+    #     input_vec = tl.view(input_vec, [num_input_groups, 1, in_group_size])
+    #     , but this does not work because tl.view may reorder elements arbitrarily; see its docstring
+
+    # Stage 2: load integer codes for the active row
+    # [in_features // in_group_size, num_codebooks]
+    codes_i_ptrs = (
+        codes_i16_ptr
+        + pid * num_input_groups * num_codebooks
+        + tl.arange(0, num_input_groups_next_power_of_2)[:, None] * num_codebooks
+        + tl.arange(0, num_codebooks)[None, :]
+    )
+    codes_i_mask_1d = tl.arange(0, num_input_groups_next_power_of_2) < num_input_groups
+
+    codes_i = tl.load(codes_i_ptrs, mask=codes_i_mask_1d[:, None])  # [in_features//in_group_size, num_codebooks]
+    if codes_i.dtype == tl.int16:
+        codes_i = codes_i.to(tl.int32)
+        codes_i = (codes_i) + (codes_i < 0) * codebook_size  # aka 2 ** nbits_per_codebook
+        # ^-- (because codes are int16 tensors that contain uint data)
+
+        # The following alternative does not work:
+        #     codes_i = codes_i.to(tl.int32) % codebook_size # aka 2 ** nbits_per_codebook
+    else:
+        codes_i = codes_i.to(tl.int32)
+
+    # shift codes_i so that codebooks after 0th point to correct indices in codebooks_ptr
+    codes_i += tl.arange(0, num_codebooks)[None, :] * codebook_size  # aka 2 ** nbits_per_codebook
+    # ^-- [in_group_size, num_codebooks]
+
+    # Stage 3: convert codes to pointers to every individual (activated) weight in codebooks
+    # [in_features // in_group_size, num_codebooks, out_group_size, in_group_size]
+    out_group_ix = tl.arange(0, out_group_size)[None, None, :, None]
+    in_group_ix = tl.arange(0, in_group_size)[None, None, None, :]
+    weight_i_ptrs = (
+        codebooks_ptr
+        + codes_i[:, :, None, None] * out_group_size * in_group_size
+        + out_group_ix * in_group_size
+        + in_group_ix
+    )
+
+    # Stage 4: reconstruct weights, multiply by inputs and write out
+    weights_i = tl.load(weight_i_ptrs, mask=codes_i_mask_1d[:, None, None, None], other=0)
+    if compute_in_fp32:
+        weights_i = weights_i.to(tl.float32)
+        input_vec = input_vec.to(tl.float32)
+    # ^-- [in_features // in_group_size, num_codebooks, out_group_size, in_group_size]
+    weights_i = tl.sum(weights_i, axis=1)  # sum codebooks as per additive quantization
```

Comment:
The code seemingly works faster if you do a single sum instead of 
Please see DM

( https://gist.github.com/justheuristic/5ef9594a7b9191db38c9d6eb71f670cf )

Reply:
I'll just remove the kernel from this PR. It would be a matter for another PR.